src/librustc_mir/interpret/snapshot.rs

```diff
@@ -0,0 +1,304 @@
+use std::hash::{Hash, Hasher};
+
+use rustc::ich::{StableHashingContext, StableHashingContextProvider};
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+use rustc_data_structures::fx::{FxHashSet, FxHasher};
+use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
+use rustc::mir;
+use rustc::mir::interpret::{AllocId, Pointer, Scalar, ScalarMaybeUndef, Value, Relocations, Allocation};
+use rustc::ty;
+use syntax::codemap;
+
+use super::eval_context::{LocalValue, StackPopCleanup};
+use super::{Frame, Memory, Machine, Place, PlaceExtra};
+
+/// The virtual machine state during const-evaluation at a given point in time.
+#[derive(Eq, PartialEq, Hash)]
+pub struct EvalSnapshot<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> {
```

Comment:
I currently can't see what this scheme gains us over the previous one. It introduces a lot of complexity.

Doesn't your suggestion of using `StableHash` as a hash with loads of collisions and then using the equality operation to look at content instead of ids work?

Reply:
That's still the idea exactly, the only thing I changed from the previous iteration was to pre-resolve `AllocId`s to something that can be compared equal by the time `EvalSnapshot` is constructed, so that that we can just derive `Eq` directly. The alternative is to unroll the equality comparison for `EvalSnapshot` manually, which turned out to be extremely complex as we saw a couple of iterations ago.

I had to create a new type for `AllocId` because I can't just reimplement `Eq`, since it has index semantics everywhere else.

Reply:
Hmm... If this level of complexity is necessary, I wonder if it's really worth doing this. We can never catch all cycles anyway.

I don't feel comfortable just adding more things to miri without a very clear benefit.

Reply:
> mapping the evaluation context to a data structure that is invariant to AllocId and may be trivially compared.

This definitely feels simpler, but it looks to me like in the end it also requires writing out all the crawling code.

Reply:
Correct, that's what the trait `Snapshot` does, maybe you have an insight on a simpler way to implement it?

Reply:
I presume we'd need both

* the HashStableContext trait ( https://github.com/rust-lang/rust/issues/53084 )
* split out all the fields that should be part of the hash/eq into their own type and thus single field so we can derive everything on them

This way the only special code would be the one for the `HashStable` impl of `AllocId`

Reply:
> split out all the fields that should be part of the hash/eq into their own type 

I'm not sure I understand your suggestion, could you elaborate?

To clarify my current approach, I created _snapshot_ versions of everything that can directly or indirectly contain an `AllocId` and the list is long: `Frame`, `Memory`, `LocalValue`, `Value`, `ScalarMaybeUndef`, `Scalar`, `Pointer`, `Allocation` and `Relocations`. Instances of this types form a [cyclic] graph that cannot easily be flattened into a simple structure, because many are enums whose variants contain other metadata besides the `AllocId`, such as _offsets_ and _alignments_, that must also be preserved and later compared for equality. The trait `Snapshot` that I introduced is very similar to the existing `HashStable` trait, except that instead of producing a hash, it clones the entire structure except for `AllocId`s, which are replaced by `AllocHashId`s.

Reply:
We would not end up comparing Memory. Instead we'd query Memory for its Allocation whenever we encounter an `AllocId`.

To support this we'd need a copy of `Memory`, so we can just clone the entire thing (let's ignore code that has memory leaks for now, as those would cause `Memory` to grow even if the amount of observable state does not grow)

Then the lhs of the comparison uses the current values and current memory object and the rhs of the comparison uses the old values and the clone of the old memory.

The EvalContext contains the Memory, but also a bunch of other things that we do not need (like the TyCtxt, the loop counters, ...) so everything but the irrelevant fields could be moved into a separate structure ("state" maybe?) and then we can clone, hash and compare these fields via their derived traits instead of writing everything by hand...

But maybe we do still need handwritten impls for equality, because comparisons have no state... so we'd end up with the Situation we have now... ugh..

Ok... so maybe your solution is the best one. I'll give it another review.

Reply:
Right, the "state" you mention is `EvalSnapshot`, it needs to contain a _snapshot_ of `Machine`, the stack (`Vec<Frame>`) and `Memory`, where _snapshot_ here means a clone of the subset of fields that matter for the purpose of detecting repeated states.

Hashing this snapshot is the easy part, since the current implementation of `HashStable` sufficient (albeit not perfect). The problem is implementing `Eq`, because we can't just implement it for `AllocId` somehow, since we need `Memory` to be accessible by the time we compare them.

Reply:
> The problem is implementing Eq, because we can't just implement it for AllocId somehow, since we need Memory to be accessible by the time we compare them.

Yes, that's where I notices why your solution might be more elegant ;)

Sorry about the confused mess above

Reply:
After thinking about the current approach versus the direct implementation of `PartialEq` for `EvalSnapshot` for a few days, I'm starting to think that despite the amount of *scary hand written code* required by the later, it's still better than the former:

1. The current approach has the serious issue that it still relies on hashing in order to implement `PartialEq`, which means it could potentially cause false positives and reject valid code. I've attempted several approaches to fix this over the past few days, but I wasn't able to come up with anything reasonable so far.
2. Since the equality comparison of `FrameSnapshot` and `MemorySnapshot` are independent of one another, we may spend precious CPU cycles comparing unreachable `Allocation`s.

Based on this I'm inclined to stash the current approach on a branch and give another shot at manually implementing `PartialEq` for `EvalSnapshot`. 

@oli-obk what do you think?

Reply:
Yes this seems sensible. Maybe you can build a helper macro similar to the ones for `HashStable` to reduce the amount of boilerplate?